Client enhancements

mbohlool opened the pull request:
- Adding kube config support (load_kube_config)
- ~~Adding SSL support (a fix for python 2.7 to accept ipaddress hosts)~~
- Adding streaming watch support with test
- Adding preprocessing step to enhance operation IDs
- Regenerated client with enhanced swagger.json and various fixes in swagger-codegen swagger-api/swagger-codegen#4173 swagger-api/swagger-codegen#4166 swagger-api/swagger-codegen#4152 and kubernetes fixes kubernetes/kubernetes#36684
- Adding examples
- Removing generated tests

fixes #4

sebgoa commented:
there is a requirement on pyyaml to be added, for the k8sutil to work on python 2.7

mbohlool commented:
Good catch sebgoa. I will add it. I will update this PR with much better commits soon. Support for watch streaming response is among them.

mbohlool commented:
@mml @smarterclayton @caesarxuchao Friendly ping.

caesarxuchao commented:
First 3 commits lgtm.

caesarxuchao commented:
The logic of the 4th commit makes sense. I'm not familiar with Python, so I can't give comments on the coding.

caesarxuchao commented on k8sutil/ssl.py:
```diff
@@ -0,0 +1,61 @@
+    try:
+        # Divergence from upstream: ipaddress can't handle byte str
+        host_ip = _ip_address(hostname)
+        san = cert.get('subjectAltName', ())
```
What is `()`?

caesarxuchao replied:
Is my understanding correct: this commit is fixing match_hostname function of python version < v3, so that if the hostname is an IP address, it can still match.

mbohlool replied:
() is the default value, it means an empty pair will return if cert does not have 'subjectAltName'.

Your understanding is correct.

caesarxuchao commented on k8sutil/watch.py:
```diff
@@ -0,0 +1,107 @@
+                        j = json.loads(l)
+                        j['raw_object'] = j['object']
+                        if return_type:
+                            obj = SimpleNamespace(data=json.dumps(j['raw_object']))
```
What does SimpleNamespace do?

mbohlool replied:
ref: https://docs.python.org/3.5/library/types.html#types.SimpleNamespace

It is something that available in python3 not python2.  it lets you have types with dynamic attributes.

marun commented on k8sutil/watch.py:
```diff
@@ -0,0 +1,107 @@
+        resp = func(*args, **kwargs)
+        try:
+            prev = ""
+            for seg in resp.read_chunked(decode_content=False):
```
Consider simplifying by breaking some of this out into a generator:

```python
def iter_resp_lines(self, resp):
  prev = ""
  for seg in resp.read_chunked(decode_content=False):
    if isinstance(seg, bytes):
      seg = seg.decode('utf8')
    seg = prev + seg
    lines = seg.split("\n")
    if not seg.endswith("\n"):
      prev = lines[-1]
      lines = lines[:-1]
    else:
      prev = ""
    for l in lines:
      if l:
        yield l
...
  try:
    for l in self.iter_resp_lines(resp):
      ...
```

mbohlool replied:
Done.

marun commented on k8sutil/watch_test.py:
```diff
@@ -0,0 +1,136 @@
+        fake_resp.done()
+        fake_api.done()
+
+    def test_watch_with_int_object(self):
```
Is this test doing more than validating ``json.dumps``?

mbohlool replied:
It is testing expected behavior of the Watch class that is a return type should be honored. I think it is not about what is testing now, it is about expected behavior and future changed not to alter that.

mbohlool replied:
and more tests never hurt :)

marun replied:
I think tests have a cost - they increase the maintenance burden - and that their addition should be justifiable rather than 'always good'.  Given that this is a unit test, and there is no special handling in the code for in objects, I don't see the point of validating behavior that is essentially the responsibility of the json library.  The json library should already have tests for this.

marun commented on k8sutil/watch_test.py:
```diff
@@ -0,0 +1,136 @@
+        fake_resp.done()
+        fake_api.done()
+
+    def test_watch_with_no_return_type(self):
```
This is an awful lot of code to validate return type detection.  Consider breaking out detection of return type from ``stream`` so that it can be validated separately from reading a response.

mbohlool replied:
I am not sure if I understand your point here. This test assumes there is no return type and therefore all types should stay the same. It just test that behavior of watch.

marun replied:
I think that validating the behavior when no return type is specified could be performed without having to call ``stream`` if the code were factored properly.  While it is possible to validate this behavior at the level of the watch, the result is a brittle test that can be broken by changes unrelated to the functionality under test (the return type handling). 

write tests at I subscribe to the idea of testing at the lowest level possible, since that minimizes the complexity that needs to be tested.  An example

marun commented on k8sutil/test_utils.py:
```diff
@@ -0,0 +1,49 @@
+    pass
+
+
+class FakeClass(object):
```
This looks an awful lot like the start of a mocking library.  Any reason not to use an existing one?  Or better yet, skip mocking and focus on writing code that can be unit tested without mocks?  It would be simpler to write and easier to maintain.

mbohlool replied:
Do you have a suggestion for a mocking lib? I searched a little and didn't find a good one and found it easier to just write a simple one.

I think in case of this Watch, having mock is not a bad idea, I can break some pieces, but the total behavior (integration test?) should also be tested and that is much easier with a Mock.

Here I didn't focus on pure unit or integration test and tried to cover all code paths as much as possible.

mbohlool replied:
Pardon my ignorance. There is an easy to use Mock library. I've removed this utility class and used it.

marun commented on k8sutil/kube_config.py:
```diff
@@ -0,0 +1,160 @@
+def _load_cluster_info(cluster, user):
+    """Loads cluster information from kubeconfig such as host and SSL certificates."""
+    if 'server' in cluster:
+        configuration.host = cluster['server']
```
Hmm, this is a global?  Will this library be capable of contacting multiple kube clusters at the same time, or would that require multiple processes?

mbohlool replied:
the generated code defined configuration as a singleton. I am not sure why is that, I would file an issue to follow up on that with swagger-codegen ppl and fix it in codegen if that make sense. I agree that this client should support connecting to multiple kube clusters.

marun commented on k8sutil/kube_config.py:
```diff
@@ -0,0 +1,160 @@
+        os.remove(f)
+
+
+def _create_temp_file_with_content(content):
```
Rather than managing many created tmp files, consider tracking a single directory created using ``mkdtemp`` and call ``mkstemp`` against that path.

mbohlool replied:
Why that is be better? I mean I am just letting os do what is best. There is no requirement for these files to be in the same folder.

marun commented on k8sutil/kube_config.py:
```diff
@@ -0,0 +1,160 @@
+        return _create_temp_file_with_content(o[data_key_name])
+    if file_key_name in o:
+        return o[file_key_name]
+    return None
```
Consider removing this line, since python returns None by default.

marun commented on k8sutil/kube_config.py:
```diff
@@ -0,0 +1,160 @@
+    return name
+
+
+def _find_object_with_name(o, name, o_name):
```
Consider avoiding the use of single-character variables for anything but loop indexes.  ``obj`` might be more appropriate here.  Also, what is the import of ``name`` vs ``o_name``?  It's non-obvious to me, and likely to be  non-obvious to future maintainers.

mbohlool replied:
This whole function was obsolete. sorry for that. removed. (name was the value of attribute 'name' of the object, and o_name was a debug friendly name of the object such as 'user' or 'cluster'. the logic moved to _node.get_with_name)

marun commented on k8sutil/kube_config.py:
```diff
@@ -0,0 +1,160 @@
+        with open(o[file_key_name], 'r') as f:
+            data = f.read()
+        return data
+    return None
```
ditto

mbohlool replied:
Done.

sebgoa commented:
Honestly this PR is so big with auto-generated code, you should just merge it and move on in master.

I think it is more important to setup build bots, pip package, proper README and a bit of doc.

Right now what is in master is broken.

marun commented on k8sutil/ssl.py:
```diff
@@ -0,0 +1,61 @@
+        pass
+    return prev_match_hostname(cert, hostname)
+
+def fix_ssl_hosts_with_ipaddress():
```
Is this really the best way to solve this problem?  Monkeypatches are risky even when the patch is in code intended to be consumed as an application (and therefore packaged with a known good version of the targeted dependency).  Patching in a library is generally a bad idea because there is no guarantee that the consumer will have a compatible patch target, and bugs in the patch may prove difficult to track down. 

If this is really the only option, consider adding an explicit warning that monkeypatching is being performed to the README to aid in troubleshooting.

mbohlool replied:
I see the ugliness in what I am doing here and I completely agree with you that is not a clean fix. I spend some time on this. basically py2.7 does not support ipaddress hosts and the fix has been added here: https://bugs.python.org/issue23239 for python 3.x. I am open to suggestions here and I will keep looking for a better solution. For now I will go with your suggestion of adding a warning to README file.

mbohlool replied:
Look like this have been fix in newer versions of urllib3. I will confirm that and remove this hack all-together in that case.

mbohlool commented on k8sutil/watch.py:
```diff
@@ -0,0 +1,107 @@
+
+def _find_return_type(func):
+    for line in pydoc.getdoc(func).splitlines():
+        if line.startswith(":return:"):
```
Done.

mbohlool commented on k8sutil/watch.py:
```diff
@@ -0,0 +1,107 @@
+            return_type = _find_return_type(func)
+            # Hacky assumption that watching `func` return TypeList in non-watch mode and an event with
+            # object of type in Type in watch mode.
+            if return_type.endswith("List"):
```
Done.

mbohlool commented on k8sutil/watch.py:
```diff
@@ -0,0 +1,107 @@
+            # Hacky assumption that watching `func` return TypeList in non-watch mode and an event with
+            # object of type in Type in watch mode.
+            if return_type.endswith("List"):
+                return_type = return_type[:len(return_type)-4]
```
Done.

mbohlool commented on k8sutil/watch.py:
```diff
@@ -0,0 +1,107 @@
+                else:
+                    prev = ""
+                for l in lines:
+                    if l:
```
Not relevant anymore because of previous refactoring.

mbohlool commented:
@sebgoa I would like to have at least one person with python background to sign off on this before merging. The auto-generated code can be ignored and the rest of the code should not be hard to review.

mbohlool commented:
@marun Thanks for the review. I've applied or replied to your comments. Notably I've removed ssl hack and changed dependency to a newer version of urllib3 that fixed the ip address hosts. Please take another look.

DirectXMan12 commented on k8sutil/watch.py:
```diff
@@ -0,0 +1,115 @@
+TYPE_LIST_SUFFIX = "List"
+
+
+class SimpleNamespace:
```
you can probably replace this with a named tuple (see `collections.namedtuple`), if you know what fields you need (and it looks like there's only one use below, so you probably can).

marun commented on k8sutil/watch.py:
```diff
@@ -0,0 +1,115 @@
+            for l in iter_resp_lines(resp):
+                j = json.loads(l)
+                j['raw_object'] = j['object']
+                if return_type:
```
(No action required) 

Consider factoring out the marshaling so it can be isolated for testing, e.g.

```python
def marshall_obj(data, return_type):
 j = json.loads(data)
 j['raw_object'] = j['object']
 if return_type:
   ...
 return j 

def wait(...):
   ...
   try:
      for l in iter_resp_lines(resp):
         obj = self.marshall_obj(l, return_type) 
         yield obj
      ...
```

mbohlool replied:
Done.

mbohlool commented on k8sutil/kube_config.py:
```diff
@@ -0,0 +1,148 @@
+# Copyright 2016 The Kubernetes Authors.
```
Will do.

mbohlool commented on k8sutil/watch.py:
```diff
@@ -0,0 +1,115 @@
+        self._stop = True
+
+    def stream(self, func, *args, **kwargs):
+        """Watch an API resource and stream the result back as a generators.
```
Done

mbohlool commented on k8sutil/watch_test.py:
```diff
@@ -0,0 +1,128 @@
+
+import unittest
+from .watch import Watch
+from mock import Mock
```
Thanks for the video. It was useful. I agree that complicated Mocks will make things harder to follow and keeping test/mock with code would be harder. My point here is the Mock is a simple request mock and does not have much logic. Maybe if I refactor the test itself, it would be more clear. Nevertheless, you refactoring suggestions are good and I am going to apply them as much as possible.

mbohlool commented on k8sutil/watch_test.py:
```diff
@@ -0,0 +1,128 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
```
Added a semi-automatic script to do these things.

mbohlool commented:
I've applied/replied to all comments and will merge this.

Thanks @marun, @sebgoa, @DirectXMan12, @caesarxuchao for review.
